jobs/activateLicences.ts

```diff
@@ -0,0 +1,88 @@
+import logger from '../logger'
+
+import generateOauthClientToken from '../server/authentication/clientCredentials'
+import config from '../server/config'
+import { LicenceSummary } from '../server/@types/licenceApiClientTypes'
+import LicenceStatus from '../server/enumeration/licenceStatus'
+import LicenceApiClient from '../server/data/licenceApiClient'
+import PrisonerSearchApiClient from '../server/data/prisonerSearchApiClient'
+import { Prisoner } from '../server/@types/prisonerSearchApiClientTypes'
+
+initialiseAppInsights()
+buildAppInsightsClient('create-and-vary-a-licence-activate-licences-job')
+
+const pollLicencesToActivate = async (): Promise<LicenceSummary[]> => {
+  const approvedLicences = await getApprovedLicences()
+  const nomisIds = approvedLicences.map(licence => licence.nomisId)
+  const prisonersWithApprovedLicences = await getPrisoners(nomisIds)
+  const releasedPrisonerNumbers = prisonersWithApprovedLicences
+    .filter(prisoner => {
+      return (
+        prisoner.status.startsWith('INACTIVE') &&
+        prisoner.confirmedReleaseDate &&
+        moment(prisoner.confirmedReleaseDate, 'YYYY-MM-DD').isSameOrBefore(moment())
+      )
+    })
+    .map(prisoner => prisoner.prisonerNumber)
+  return approvedLicences.filter(licence => releasedPrisonerNumbers.includes(licence.nomisId))
+}
+
+const getSystemClientTokenFromHmppsAuth = (): Promise<superagent.Response> => {
+  const clientToken = generateOauthClientToken(
+    config.apis.hmppsAuth.systemClientId,
+    config.apis.hmppsAuth.systemClientSecret
+  )
+
+  const authRequest = querystring.stringify({ grant_type: 'client_credentials' })
+
+  logger.info(`HMPPS Auth request '${authRequest}' for client id '${config.apis.hmppsAuth.systemClientId}'`)
+
+  return superagent
+    .post(`${config.apis.hmppsAuth.url}/oauth/token`)
+    .set('Authorization', clientToken)
+    .set('content-type', 'application/x-www-form-urlencoded')
+    .send(authRequest)
+    .timeout(config.apis.hmppsAuth.timeout)
+}
+
+const getApprovedLicences = async (): Promise<LicenceSummary[]> => {
+  const authResponse = await getSystemClientTokenFromHmppsAuth()
+  return new LicenceApiClient(authResponse.body.access_token).matchLicences([LicenceStatus.APPROVED])
+}
+
+const getPrisoners = async (nomisIds: string[]): Promise<Prisoner[]> => {
+  if (nomisIds.length === 0) {
+    return []
+  }
+
+  const authResponse = await getSystemClientTokenFromHmppsAuth()
+  const prisonerSearchCriteria = {
+    prisonerNumbers: nomisIds,
+  }
+  return new PrisonerSearchApiClient(authResponse.body.access_token).searchPrisonersByNomsIds(prisonerSearchCriteria)
+}
+
+const batchActivateLicences = async (licenceIds: number[]): Promise<void> => {
+  if (licenceIds.length > 0) {
+    const authResponse = await getSystemClientTokenFromHmppsAuth()
+    await new LicenceApiClient(authResponse.body.access_token).batchActivateLicences(licenceIds)
+  }
+}
```

Comment:
What happens on a ValidationException? Caught and a 400 returned here?  I think it unlikely to happen between getting the list and calling the API, so maybe take out the ValidationException check in the API? Sorry, I know it was my suggestion but it complicates things and probably would never happen - particularly if the job runs early each morning.

Reply:
I've since read the whole thing, and its being caught. But still think its probably something that wouldn't happen, and if it did, just skipping that one licence would be better than failing the job?

Reply:
Yep it will bomb out.. I will remove the exception from the API and just log it instead